core-v1/src/main/scala/lens/VariantLens.scala

```diff
@@ -0,0 +1,269 @@
+    */
+  def id[A]: VariantLens[A, A] = Id.asInstanceOf[Id[A]]
+
+  /**
+    * A function for building a lens from the [[Id]] lens.
+    */
+  type Fn[-A, +B] = VariantLens[Any, A] => VariantLens[A, B]
+
+  implicit object ComposeInstance extends Compose[VariantLens] {
+    override def compose[A, B, C](
+      f: VariantLens[B, C],
+      g: VariantLens[A, B],
+    ): VariantLens[A, C] =
+      g.andThen(f)
+  }
+
+  /**
+    * A fancy trick using macros that allows you to access the surrounding context of a method
+    * invocation. By wrapping the call with this class, we can access the given [[lens]] to
+    * apply the appropriate transformation.
+    *
+    * @param lens the lens to add the `.select` method to.
+    */
+  implicit final class Selector[A, B](val lens: VariantLens[A, B]) extends AnyVal {
+
+    /**
+      * Create a [[VariantLens]] that selects a field based on a given function and applies the appropriate
+      * [[DataPath.atField]] operator.
+      *
+      * @note this uses a macro and only works for `val`s or `def`s with no arguments.
+      *
+      * @param getter a function that selects a value of type [[C]] from the return type [[B]] of the current lens
+      */
+    def select[C](getter: B => C): VariantLens[A, C] = macro VariantLensMacros.selectImpl[A, B, C]
+  }
+
+  /**
+    * If the lens returns a [[IterableOnce]] of 2-tuples, then allow calling operations defined by [[AsMapBuilder]]
+    */
+  implicit def asMap[A, K, V](lens: VariantLens[A, IterableOnce[(K, V)]]): AsMapBuilder[A, K, V] =
+    new AsMapBuilder[A, K, V](lens)
+
+  final class AsMapBuilder[A, K, V](private val lens: VariantLens[A, IterableOnce[(K, V)]]) extends AnyVal {
+
+    /**
+      * Returns the right-side of all the 2-tuples as an [[Iterable]] of values.
+      */
+    def values: VariantLens[A, Iterable[V]] = lens.copy(
+      get = lens.get.andThen(_.iterator.to(View).map(_._2)),
+    )
+
+    /**
+      * @see [[AsIterableBuilder.to]] but the values are 2-tuples
+      */
+    def to[B](factory: Factory[(K, V), B]): VariantLens[A, B] =
+      lens.copy(get = lens.get.andThen(factory.fromSpecific(_)))
+
+    /**
+      * Group the 2-tuples by the left side and create a [[Map]].
+      *
+      * If there are duplicate keys, then the last entry in the sequence wins the spot and the others are dropped.
+      */
+    def toMap: VariantLens[A, Map[K, V]] = lens.copy(
+      get = lens.get.andThen(Map.from(_)),
+    )
+
+    /**
+      * Same as [[toMap]], but creates a lazy [[MapView]] instead.
+      */
+    def toMapView: VariantLens[A, MapView[K, V]] = lens.copy(
+      get = lens.get.andThen(Map.from(_).view),
+    )
+  }
+
+  // TODO: Make this more generally useful and convert back to a NamedLens implicitly
+  // I think this can be folded into VariantLens now that it utilizes variance
+  final class AsIterableBuilder[A, C[x] <: IterableOnce[x], E](private val lens: VariantLens[A, C[E]]) extends AnyVal {
```

Comment:
I think this TODO is possible, but I don't have time to experiment with this for the first milestone release.